test-suite/test-functionality.sh

```diff
@@ -0,0 +1,264 @@
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+# This script tests a few Squid functionality requirements.
+# It is suitable for automated CI and manual testing.
+
+# Default-set and report used environment variables:
+# * the root directory for storing test tools and test artifacts.
+echo "TMPDIR=${TMPDIR:=${RUNNER_TEMP:-/tmp}}"
+# * directory for cloning git repositories containing various test tools
+echo "CLONES_DIR=${CLONES_DIR:=$TMPDIR/clones}"
+# * directories of cloned repositories
+echo "DAFT_DIR=${DAFT_DIR:=$CLONES_DIR/daft}"
+echo "SQUID_DAFTS_DIR=${SQUID_DAFTS_DIR:=$CLONES_DIR/squid-dafts}"
+echo "SQUID_OVERLORD_DIR=${SQUID_OVERLORD_DIR:=$CLONES_DIR/squid-overlord}"
+
+# print an error message (with special markers recognized by Github Actions)
+echo_error() {
+    echo "::error ::" "$@"
+}
+
+# print a warning message (with special markers recognized by Github Actions)
+echo_warning() {
+    echo "::warning ::" "$@"
+}
+
+run() {
+    echo "running: $@"
+    "$@"
+}
+
+# Whether the branch has a commit with a message matching the given regex.
+# The first argument is the SHA of the primary commit with the desired change.
+# We want to find up/backported commits that will have different SHAs and that
+# may have minor commit message variations, so that SHA is for reference only.
+has_commit_by_message() {
+    local commit="$1"
+    shift
+
+    if git log --grep "$@" | grep -q ^commit
+    then
+        return 0;
+    fi
+
+    echo "Was looking for branch commit with a message matching the following"
+    echo "    regex: " "$@"
+    echo "    This code lacks commit $commit or equivalent."
+    return 1;
+}
+
+clone_repo() {
+    local repo_url="$1"
+    local destination_dir="$2"
+
+    if test -e $destination_dir
+    then
+        echo "Skipping already fetched $destination_dir"
+    elif run git clone --no-tags --quiet --depth=1 --branch production -- "$repo_url" "$destination_dir"
+    then
+        if test -e "$destination_dir/package.json"
+        then
+            ( cd $destination_dir && run npm install --no-audit --no-save ) || return
+        fi
+    else
+        echo_error "Failed to fetch $repo_url into $destination_dir"
+        return 1
+    fi
+
+    (cd $destination_dir && echo "Using commit `git rev-parse HEAD`")
+}
+
+start_overlord() {
+    local url=http://localhost:13128/
+    local log=$TMPDIR/squid-overlord.log
+    if test -e $log && curl -H 'Pop-Version: 4' --no-progress-meter $url/check > /dev/null
+    then
+        echo "Will reuse squid-overlord service running at $url"
+        return 0;
+    fi
+
+    # Do not be tempted to simply run `sudo ... overlord.pl`: User nobody will
+    # lack read permissions, and sudo will ask for a password.
+    sudo -n --background -u nobody perl < $SQUID_OVERLORD_DIR/overlord.pl > $log 2>&1 || return
```

Comment:
This is the only `sudo` left in the test scripts. I do not see a good way of removing it because the overlord.pl process should run as the Squid user (i.e. `nobody`) and these scripts usually run as a Github Actions user or, when executed manually, as a developer.

This `sudo` is not a problem for Actions (it does not ask for a password there).

If this `sudo` step bothers a developer that runs this script manually and has a password-asking sudo setup, then they do have an option of running this command before starting the manual test -- the script will then detect the overlord.pl process running and skip this sudo step. The same trick is useful if one wants to run a custom overlord.pl version (which is useful during test case development). This trick is one of the reasons why the script does not try to kill the overlord.pl process at the end of the test.

Reply:
FYI, there are some tricks I have picked up from Debian people regarding solutions to safely kill the script on closure. We likely may even be able to use the Debian/Ubuntu developer tools instead of sudo. That is a change for later though.